Porting stock move backdating

archetipo opened the pull request.

gurneyalex committed:
Update travis configuration

use sudo: false to enable container build
use cache: pip to cache pip packages

archetipo committed:
[FIX] porting of the module
[FIX] tests
[FIX] translation

pedrobaeza commented:
I'm not sure that with new quant approach this module is still valid. Have you tried?

archetipo commented:
Hi @pedrobaeza 
yes i've tried.. why you are not sure?

pedrobaeza commented:
Because quants are created always in the current date, so there will be an incongruency. It will also can create a paradox like the one of the fictional movies (how can you delivery things that don't exist because they have been already delivered in the future?)

archetipo commented:
@pedrobaeza   ok Thanks

archetipo committed:
[FIX] update quants date
[FIX] test

eLBati committed:
FIX when using module with user != admin. Test added

eLBati commented:
@archetipo please see https://github.com/archetipo/stock-logistics-workflow/pull/1

tafaRU commented on stock_move_backdating/README.rst:
```diff
@@ -0,0 +1,45 @@
+    :alt: License: AGPL-3
+
+Allows back-dating of stock moves
+===========
```
It has to be to  “underlined” using equal signs.

gurneyalex commented:
needs rebase

archetipo committed:
[FIX] README
[FIX] Travis.yaml
[FIX] rebased with 8.0

archetipo commented:
Thanks @gurneyalex

eLBati commented on .travis.yml:
```diff
@@ -1,10 +1,5 @@
-  apt: true
-  directories:
-    - $HOME/.pip-cache/
+cache: pip
```
@archetipo probably travis is broken due this change. Why is it there?

blaggacao replied:
I wondered, too.

blaggacao commented on .travis.yml:
```diff
@@ -1,10 +1,5 @@
-language: python
```
Ignorant question: Is this sensible?

eLBati replied:
MQT scripts are python: https://github.com/OCA/maintainer-quality-tools/tree/master/travis

In case of doubts, I would just follow the sample file
https://github.com/OCA/maintainer-quality-tools/blob/master/sample_files/.travis.yml

blaggacao replied:
so, your answer is no, it is not sensible to remove it. But for whatever reason it is removed, we don't know if @archetipo  has done it on purpose.

archetipo commented:
@eLBati  @blaggacao Thnaks! Travis.yaml was fixed

eLBati commented on stock_move_backdating/__openerp__.py:
```diff
@@ -19,23 +19,18 @@
+    'name': 'Stock Move Backdating',
+    'version': '1.0',
+    'category': 'Stock Logistics',
+    'author': 'Bremskerl, Odoo Community Association (OCA)',
```
Why did you change `author`?

archetipo replied:
author is here https://github.com/OCA/stock-logistics-workflow/pull/142/files#diff-8ebcf17df81968ac09b007cf5d372958R28

eLBati replied:
@archetipo that's not a good reason I think :bowtie:

archetipo committed:
[FIX] author

archetipo commented:
Thanks @eLBati  your remarks were fixed.

eLBati commented:
@archetipo please also move the module to https://github.com/archetipo/stock-logistics-workflow/blob/porting_stock_move_backdating/README.md#available-addons

gurneyalex commented on stock_move_backdating/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,42 @@
+    @api.constrains('date')
+    def check_date(self):
+        now = datetime.now().strftime(
+            DEFAULT_SERVER_DATETIME_FORMAT)
```
I think there is a TZ mismatch here, which will cause a bug:  `self.date` is in UTC, while `now` is in the local TZ of the server. I think you need to use `datetime.utcnow()` to fix this.

archetipo replied:
@gurneyalex  ok Thanks

gurneyalex commented on stock_move_backdating/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,42 @@
+    _inherit = 'stock.transfer_details_items'
+
+    _defaults = {
+        'date': datetime.now(),
```
please change this to `fields.Datetime.now` (without parenthesis) (recommended function for default value)

gurneyalex commented on stock_move_backdating/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,42 @@
+    @api.one
+    @api.constrains('date')
+    def check_date(self):
+        now = datetime.now().strftime(
```
please change the call to `strftime` to a call to `fields.Datetime.to_string`

gurneyalex commented on stock_move_backdating/model/stock.py:
```diff
@@ -0,0 +1,63 @@
+            if move.linked_move_operation_ids:
+                operation = move.linked_move_operation_ids[0]
+                move.date = operation.operation_id.date
+                move.refresh()
```
I don't think this call is necessary (and if it is not, it will probably badly harm the performances). Can you check?

gurneyalex commented on stock_move_backdating/model/stock.py:
```diff
@@ -0,0 +1,63 @@
+                move.refresh()
+                if move.quant_ids:
+                    for quant in move.quant_ids:
+                        quant.sudo().in_date = move.date
```
since you are writing the same date on all the quants, please do this in a single call to write, rather than in a loop.

gurneyalex commented on stock_move_backdating/model/stock.py:
```diff
@@ -0,0 +1,63 @@
+        res = super(stock_quant, self)._create_account_move_line(
+            quants, move, credit_account_id, debit_account_id, journal_id
+        )
+        for o2m_tuple in res:
```
This method has no automated test (see https://coveralls.io/builds/4018455/source?filename=stock_move_backdating%2Fmodel%2Fstock.py)

AFAICT, `stock_quant._create_account_move_line` returns `None` (see https://github.com/odoo/odoo/blob/8.0/addons/stock_account/stock_account.py#L237) so I expect this code to crash with None not being iterable. 

Please add an automated test exercising the method and fix the crash :smile_cat:

gurneyalex commented:
@archetipo code review found a :bug:

archetipo committed:
[FIX] Wizard date comparison
[RM]  useless inheritance class stock_quant and  method _create_account_move_line
[FIX] upgrade quant_ids date in a single call

archetipo committed:
[FIX] stock logistic workflow available addons

archetipo commented:
@gurneyalex  @eLBati  thanks for review, all remarks are been fixed

tafaRU commented on README.md:
```diff
@@ -36,6 +36,7 @@ addon | version | summary
 [stock_picking_reorder_lines](stock_picking_reorder_lines/) | 8.0.0.1.0 | Provide a new field on stock moves, allowing to manage the orders of moves in a picking.
 [stock_split_picking](stock_split_picking/) | 8.0.1.0.0 | Split a picking in two unconfirmed pickings
 [stock_transfer_split_multi](stock_transfer_split_multi/) | 8.0.1.0.0 | In the stock transfer wizard, you can split by multiple units
+[stock_move_backdating](stock_move_backdating/) | 8.0.0.1.0 | Allows back-dating of stock moves
```
It's not necessary, it's done automatically.

tafaRU commented on stock_move_backdating/__openerp__.py:
```diff
@@ -19,23 +19,19 @@
-    'author': ['Marco Dieckhoff, BREMSKERL', 'Agile Business Group'],
-    "category": "Stock Logistics",
+    'name': 'Stock Move Backdating',
+    'version': '8.0.0.1.0',
```
It should be **8.0.1.0.0**

archetipo replied:
here seems be possible to use also 8.0.0.1.0
https://github.com/archetipo/stock-logistics-workflow/blob/porting_stock_move_backdating/README.md#available-addons

archetipo committed:
[FIX] travis remarks

archetipo committed:
[FIX] version number

eLBati commented:
@archetipo please see commits made in august: they have wrong committer, like

```
$ git cat-file commit 7b7039fc6f45958949a8c4d760262bdc1211a6aa
tree fbd6ddcb413a6d7cc9e60ef34d5e8582e25ce95f
parent 6667cdcfcb9bd45bf4303552efd25982e24efb93
author Alexandre Fayolle <alexandre.fayolle@camptocamp.com> 1437771965 +0200
committer archetipo <archetipo@localhost> 1445587634 +0200
```

Rebase should fix them

archetipo commented:
@eLBati  ok thnaks!

archetipo commented:
@eLBati  rebase done

tafaRU commented on stock_move_backdating/__init__.py:
```diff
@@ -3,7 +3,7 @@
 #    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
 #    Author Marco Dieckhoff
-#    Copyright (C) 2013 Agile Business Group sagl (<http://www.agilebg.com>)
+#    Copyright (C) 2013 Agile Business Group (<http://www.agilebg.com>)
```
Please update copyright year, you should set as `2013-15`

tafaRU commented on stock_move_backdating/model/__init__.py:
```diff
@@ -0,0 +1,26 @@
+#
+#    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
+#    Author Marco Dieckhoff
+#    Copyright (C) 2013 Agile Business Group (<http://www.agilebg.com>)
```
Same as above.

tafaRU commented on stock_move_backdating/model/stock.py:
```diff
@@ -0,0 +1,41 @@
+#
+#    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
+#    Author Marco Dieckhoff
+#    Copyright (C) 2013 Agile Business Group sagl (<http://www.agilebg.com>)
```
Same as above.

tafaRU replied:
For consistency, please remove sagl at the end of the author name ;-)

tafaRU commented on stock_move_backdating/tests/__init__.py:
```diff
@@ -0,0 +1,24 @@
+#
+#    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
+#    Author Marco Dieckhoff
+#    Copyright (C) 2013 Agile Business Group  (<http://www.agilebg.com>)
```
Same as above.

tafaRU commented on stock_move_backdating/wizard/__init__.py:
```diff
@@ -1,7 +1,7 @@
 ##############################################################################
 #
-#    Copyright (C) 2013 Agile Business Group sagl (<http://www.agilebg.com>)
+#    Copyright (C) 2013 Agile Business Group (<http://www.agilebg.com>)
```
Same as above.

tafaRU commented on stock_move_backdating/__openerp__.py:
```diff
@@ -19,23 +19,19 @@
-    (with date != now).
-    On stock moves, user can specify the "Actual Movement Date", that will be
-    used as movement date""",
+    'depends': ['stock_account'],
```
Why did you change it?

archetipo replied:
@tafaRU 
because from version 8.0 stock.move is not  in stock but stock.account

pedrobaeza replied:
stock.move is defined in _stock_. Module _stock_account_ is a link for invoicing pickings only.

archetipo replied:
It says Pedro , my mistake , the change need to import both stock that account , with one module useful to have all dependencies met , and then have the link with the billing . Sorry but I wrote the module 4 months ago ...

tafaRU commented:
@archetipo, could you also please use simple headers following OCA template
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L1-L3 ?

Thank you in advance.

archetipo committed:
[FIX] Copyright in py module

archetipo commented:
@tafaRU 
thanks for review, done

lepistone commented:
:+1: thanks!

tafaRU commented on stock_move_backdating/tests/test_stock_move_backdating.py:
```diff
@@ -0,0 +1,114 @@
+        self.picking_model = self.env["stock.picking"]
+        self.product_model = self.env["product.product"]
+        self.location_model = self.env["stock.location"]
+        self.stock_transfer_details = self.registry("stock.transfer_details")
```
Here, too, it would be better to use `self.env` instead of `self.registry`.

archetipo commented:
@tafaRU  thanks done

tafaRU commented:
@archetipo, some Travis build jobs failed, could you please check them?

archetipo commented:
@tafaRU  thanks done

eLBati commented:
:+1: thanks

tafaRU commented:
@archetipo, please see https://github.com/archetipo/stock-logistics-workflow/pull/2

tafaRU commented:
:+1:
